Consider RefSafetyRules with UnscopedRef (#75930)
dotnet/roslyn: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md

    }
}
```

> ## `UnscopedRefAttribute` cannot be used with old ref safety rules

Reviewer: What happens if we use `[InlineArray]` in an older language version? I would expect that we have symetry between that behavior and this behavior. Basically compiler should have a uniform policy of how we approach seeing attributes in older language versions (unless there are specific reasons to deviate). Pretty our policy is to ignore but want to double check. 

Can we also update the general compiler documentation / feature check list to call out what behavior we want for attributes in this context?

Reviewer 2: Ultimately it seems to better serve readers and writers of code to disallow use of `[UnscopedRef]` in cases where it is not having any effect. We are already doing this in several scenarios with latest LangVersion/TFM, where applying the attribute isn't defined to do something meaningful.

Reviewer: The other part I'm interested in though is where do we document how attributes related to language features should be treated? This is a topic that's come up several times in the last few releases. Want us to find a place to codify what the default approach should be for such attributes. It also seems like an item that should be added to the test plan.

Author: Opened https://github.com/dotnet/roslyn/pull/76112.

Wanted to use `CallerArgumentExpressionAttribute` as an example, but turns out we don't ignore it in old LangVersions (probably not that big of an issue for that particular attribute).

***Introduced in Visual Studio 2022 version 17.13***

> The `UnscopedRefAttribute` unintentionally affected code compiled by new Roslyn versions

Reviewer: ```suggestion
The `UnscopedRefAttribute` unintentionally affected code compiled by new Roslyn compiler versions
```

> The `UnscopedRefAttribute` accidentally affected code compiled by new Roslyn versions
> even when the code was compiled in the context of the old ref safety rules
> (i.e., targeting C# 10 or older with net6.0 or older).
> However, the attribute should not have an effect in that context, and that is now fixed.

Reviewer 3: Minor suggestions:
```suggestion
The `UnscopedRefAttribute` unintentionally affected code compiled by new Roslyn versions
even when the code was compiled in the context of the earlier ref safety rules
(i.e., targeting C# 10 or earlier with net6.0 or earlier).
However, the attribute should not have an effect in that context, and that is now fixed.
```

Code that previously did not report any errors in C# 10 or older with net6.0 or older can now fail to compile:

```cs
using System.Diagnostics.CodeAnalysis;
struct S
{
    public int F;

>     // previously accidentally allowed in C# 10 with net6.0
>     // now fails with the same error as if the [UnscopedRef] wasn't there
>     // (struct members cannot return 'this' or other instance members by reference)

Reviewer 3: ```suggestion
    // previously allowed in C# 10 with net6.0
    // now fails with the same error as if the [UnscopedRef] wasn't there:
    // error CS8170: Struct members cannot return 'this' or other instance members by reference
```

    [UnscopedRef] public ref int Ref() => ref F;
}
```

> To prevent misunderstanding (thinking the attribute has an effect
> but it actually does not because your code is compiled with the old ref safety rules),
> an error is reported when the attribute is used in C# 10 or older with net6.0 or older:

Reviewer 3: ```suggestion
To prevent misunderstanding (thinking the attribute has an effect
but it actually does not because your code is compiled with the earlier ref safety rules),
an error is reported when the attribute is used in C# 10 or earlier with net6.0 or earlier:
```

```cs
using System.Diagnostics.CodeAnalysis;
struct S
{
    // both are errors in C# 10 with net6.0:

>     // UnscopedRefAttribute cannot be applied to this parameter because it is unscoped by default.

Reviewer 3: ```suggestion
    // error CS9063: UnscopedRefAttribute cannot be applied to this parameter because it is unscoped by default.
```
